source/interface/azure_iot_mqtt.h

```diff
@@ -0,0 +1,380 @@
+    eAzureIoTMQTTNoDataAvailable, /** No data available from the transport interface. */
+    eAzureIoTMQTTIllegalState,    /** An illegal state in the state record. */
+    eAzureIoTMQTTStateCollision,  /** A collision with an existing state record entry. */
+    eAzureIoTMQTTKeepAliveTimeout /** Timeout while waiting for PINGRESP. */
+} AzureIoTMQTTResult_t;
+
+typedef struct AzureIoTMQTTConnectInfo
+{
+    /**
+     * @brief Whether to establish a new, clean session or resume a previous session.
+     */
+    bool xCleanSession;
+
+    /**
+     * @brief MQTT keep alive period.
+     */
+    uint16_t usKeepAliveSeconds;
+
+    /**
+     * @brief MQTT client identifier. Must be unique per client.
+     */
+    const char * pcClientIdentifier;
+
+    /**
+     * @brief Length of the client identifier.
+     */
+    uint16_t usClientIdentifierLength;
+
+    /**
+     * @brief MQTT user name. Set to NULL if not used.
+     */
+    const char * pcUserName;
+
+    /**
+     * @brief Length of MQTT user name. Set to 0 if not used.
+     */
+    uint16_t usUserNameLength;
+
+    /**
+     * @brief MQTT password. Set to NULL if not used.
+     */
+    const char * pcPassword;
+
+    /**
+     * @brief Length of MQTT password. Set to 0 if not used.
+     */
+    uint16_t usPasswordLength;
+} AzureIoTMQTTConnectInfo_t;
+
+typedef struct AzureIoTMQTTSubscribeInfo
+{
+    /**
+     * @brief Quality of Service for subscription.
+     */
+    AzureIoTMQTTQoS_t xQOS;
+
+    /**
+     * @brief Topic filter to subscribe to.
+     */
+    const char * pcTopicFilter;
+
+    /**
+     * @brief Length of subscription topic filter.
+     */
+    uint16_t usTopicFilterLength;
+} AzureIoTMQTTSubscribeInfo_t;
+
+typedef struct AzureIoTMQTTPublishInfo
+{
+    /**
+     * @brief Quality of Service for message.
+     */
+    AzureIoTMQTTQoS_t xQOS;
+
+    /**
+     * @brief Whether this is a retained message.
+     */
+    bool xRetain;
+
+    /**
+     * @brief Whether this is a duplicate publish message.
+     */
+    bool xDup;
+
+    /**
+     * @brief Topic name on which the message is published.
+     */
+    const char * pcTopicName;
+
+    /**
+     * @brief Length of topic name.
+     */
+    uint16_t usTopicNameLength;
+
+    /**
+     * @brief Message payload.
+     */
+    const void * pvPayload;
+
+    /**
+     * @brief Message payload length.
+     */
+    size_t xPayloadLength;
+} AzureIoTMQTTPublishInfo_t;
+
+typedef struct AzureIoTMQTTDeserializedInfo
+{
+    uint16_t usPacketIdentifier;                 /** @brief Packet ID of deserialized packet. */
+    AzureIoTMQTTPublishInfo_t * pxPublishInfo;   /** @brief Pointer to deserialized publish info. */
+    AzureIoTMQTTResult_t xDeserializationResult; /** @brief Return code of deserialization. */
+} AzureIoTMQTTDeserializedInfo_t;
+
+typedef struct AzureIoTMQTTPacketInfo
+{
+    /**
+     * @brief Type of incoming MQTT packet.
+     */
+    uint8_t ucType;
```

Comment:
would be great if this was an enum

Reply:
then we will waste 3 bytes as enum will map to int32_t

---
In reply to: [598025025](https://github.com/Azure/azure-iot-middleware-freertos/pull/44#discussion_r598025025) [](ancestors = 598025025)